negcache: multiple fixes

sumit-bose opened the pull request:
- make sure domain config does not leak into global
- make sure short names are added to sub-domains
- do not use default_domain_suffix

Resolves: https://github.com/SSSD/sssd/issues/5238

alexey-tikhonov commented on src/util/util.h:
```diff
@@ -565,7 +565,11 @@ struct sss_domain_info *get_domains_head(struct sss_domain_info *domain);
 
 #define SSS_GND_DESCEND 0x01
 #define SSS_GND_INCLUDE_DISABLED 0x02
+/* Descend to sub-domains of current domain but do not go to next parent,  */
```
trailing ","

sumit-bose replied:
fixed

alexey-tikhonov commented on src/tests/cmocka/test_utils.c:
```diff
@@ -877,6 +877,37 @@ static void test_get_next_domain_flags(void **state)
     assert_null(dom);
+
+    /* Descend only to subdomains */
+    gnd_flags = SSS_GND_SUBDOMAINS | SSS_GND_INCLUDE_DISABLED;
```
Why not `SSS_GND_ALL_SUBDOMAINS`?

sumit-bose replied:
I think being explicit here helps better to understand what is going on.

alexey-tikhonov replied:
Ok

alexey-tikhonov commented on src/util/domain_info_utils.c:
```diff
@@ -39,16 +39,20 @@ struct sss_domain_info *get_next_domain(struct sss_domain_info *domain,
+            dom = dom->next;
+        } else if (descend && !only_subdomains && IS_SUBDOMAIN(dom)
+                            && dom->parent->next) {
             dom = dom->parent->next;
```
What is the meaning behind `SSS_GND_DESCEND`? Why do we go "up" beind asked to "descend"?

alexey-tikhonov replied:
Ah...
no flags == get all top-level domains **or** all sub-domain of given parent domain only
descend ==  get **all** domains
subdomains == all sub-domains of given parent only
  --  correct?

sumit-bose replied:
Yes, originally we only had top-level domains and 'descend' was added to check sub-domains as well.

alexey-tikhonov commented:
https://github.com/sumit-bose/sssd/blob/f095a01b804d97a46149c9f6593e7e99a63b7195/src/responder/common/negcache.c#L1009
  --  I think code should `continue` in case of any error, including `EAGAIN`
(there are four calls to `sss_parse_name_for_domains()` in this function)

alexey-tikhonov commented on src/tests/cmocka/test_negcache.c:
```diff
@@ -921,6 +924,255 @@ static void test_sss_ncache_reset_prepopulate(void **state)
+ *    sub-domains
+ *  - none of the names is added to a different parent domain
+ *
+ * The following table should illustrated the expextations:
```
mistype - "expextations"

sumit-bose replied:
fixed

alexey-tikhonov commented on src/tests/cmocka/test_negcache.c:
```diff
@@ -921,6 +924,255 @@ static void test_sss_ncache_reset_prepopulate(void **state)
+ * cache is repopulated each time.
+ *
+ * With the given domains, users and group we have to following expectations:
+ *  - the short name entry will be added to the domain and all sub-domains
```
I either don't understand this or can't find corresponding code.
What I see in the code: `sss_ncache_set_user()` is always called with "fqname", i.e. "name expanded to the corresponding fully-qualified name" (IIUC)

alexey-tikhonov replied:
Or should this read: ...   added **as upn**?

sumit-bose replied:
see next comment

alexey-tikhonov commented on src/tests/cmocka/test_negcache.c:
```diff
@@ -921,6 +924,255 @@ static void test_sss_ncache_reset_prepopulate(void **state)
+ * With the given domains, users and group we have to following expectations:
+ *  - the short name entry will be added to the domain and all sub-domains
+ *  - the short name expanded to the corresponding fully-qualified name of the
+ *    given domain is added to the domain and all sub-domains
```
Again, I'm not sure if I read this correctly but what I see in the code is that "fqname" is built for every domain specifically (i.e. it's not "name expanded to ... given domain is added to ... all sub-domains"

alexey-tikhonov replied:
Also those two items doesn't mention "short name as upn".
IIUC expanded short name is added as upn to domain and all sub-domains,

sumit-bose replied:
You are correct with 'What I see in the code' and my intention was to express the same in the comment. I rephrased the comment and combined the first two items into one since they meant the same.

alexey-tikhonov replied:
Thank you.

alexey-tikhonov commented on src/tests/cmocka/test_negcache.c:
```diff
@@ -921,6 +924,255 @@ static void test_sss_ncache_reset_prepopulate(void **state)
+ *  - the short name expanded to the corresponding fully-qualified name of the
+ *    given domain is added to the domain and all sub-domains
+ *  - the fully-qualified name from the parent domain is added as name to the
+ *    parent domain and as upn to all sub-domains
```
I guess this should read: ... as name **and upn** to the parent domain...?

sumit-bose replied:
fixed

alexey-tikhonov commented on src/tests/cmocka/test_negcache.c:
```diff
@@ -921,6 +924,255 @@ static void test_sss_ncache_reset_prepopulate(void **state)
+ *  - the fully-qualified name from the parent domain is added as name to the
+ *    parent domain and as upn to all sub-domains
+ *  - the fully-qualified name from the sub-domain is added as name to the
+ *    sub-domain and as upn to the parent and other sub-domains
```
The same: ... as upn to the parent and **all** sub-domains?

sumit-bose replied:
fixed

alexey-tikhonov commented:
Ouch... this stuff appeared to be surprisingly convoluted.

Nonetheless, I think patches are doing right things and should fix both issues seen in https://bugzilla.redhat.com/show_bug.cgi?id=1828483#c4 : short name treated as fqn resulting in unexpanded upn entries and missed name entries
(didn't yet test manually though).

With exception of one minor issue and a couple of mistypes, most of my comments are merely questions to make sure I got things right or requests to clarify comments.

alexey-tikhonov commented:
> Nonetheless, I think patches are doing right things and should fix both issues seen in https://bugzilla.redhat.com/show_bug.cgi?id=1828483#c4 : short name treated as fqn resulting in unexpanded upn entries and missed name entries
> (didn't yet test manually though).

Sorry it took that long. I finally verified  that those patches resolves both issues seen in https://bugzilla.redhat.com/show_bug.cgi?id=1828483#c4

Covscan is clear.

> With exception of one minor issue and a couple of mistypes, most of my comments are merely questions to make sure I got things right or requests to clarify comments.

Marking as "changes requested" to deal with this ^^.

sumit-bose commented:
> https://github.com/sumit-bose/sssd/blob/f095a01b804d97a46149c9f6593e7e99a63b7195/src/responder/common/negcache.c#L1009
> -- I think code should `continue` in case of any error, including `EAGAIN`
> (there are 3 similar spots in this function)

Hi,

this missing `continue` for the users is in fact correct and in the following code the name is treated a generic UPN. I changed the debug message to make this more clear.

For groups there are no UPNs should here we should continue on all errors. I fixed the code accordingly and comments.

bye,
Sumit

sumit-bose commented:
Hi,

thanks for the rigid review, the latest version should address your comments.

bye,
Sumit

alexey-tikhonov commented:
Thank you, ACK.

pbrezina commented:
Pushed PR: https://github.com/SSSD/sssd/pull/5361

* `master`
    * fa4b46e7de7297da3c0e37913eab8cba7f103629 - negcache: do not use default_domain_suffix
    * 0dc81a52e2836010974e9f71b1f3e47c20fd498d - negcache: make sure short names are added to sub-domains
    * 385af99ff4d5a75d0c1edc9ad830da3eb7478295 - utils: add SSS_GND_SUBDOMAINS flag for get_next_domain()
    * 0e1bcf77bd73baa0fea64830eb1f4f65a63c7afe - negcache: make sure domain config does not leak into global
* `sssd-1-16`
    * 56d509ad3001101f04c4af050c3da7472032e4cb - negcache: do not use default_domain_suffix
    * c3207deee7411456827e69d0b72d7d44e7458853 - negcache: make sure short names are added to sub-domains
    * ea32d0eb61336858aa23697b4e91d420481fc3e2 - utils: add SSS_GND_SUBDOMAINS flag for get_next_domain()
    * 96bdcbb4441ddf05c065bbafa88e5691300424d1 - negcache: make sure domain config does not leak into global
